### DISCUSSION TITLE
.Net OData - Manually get metadata entity url

### DISCUSSION THREAD

**Discussion by perci994** [OData/AspNetCoreOData#1149] [Q&A]
Hi

I have a question, I'm using OData and I'm doing it manually. That is, I am using ODataQueryOptions as a parameter and then the ApplyTo method. However, I would like to add a metadata reference as well I mean - https://localhost:5001/api/odata/$metadata#Order. Is there any way to do this please? I can't seem to find it.

```
public virtual SomeDto Handle(ODataQueryOptions<Order> query)
{
    var result = new SomeDto();

    // result.ContextUrl = request.....?; //https://localhost:5001/api/odata/$metadata#Order
    result.Items = request.Query.ApplyTo(dbQuery).ToList();

    return result;
}
```

Thanks a lot

**Comment by ElizabethOkerio** [comment]
https://github.com/OData/AspNetCoreOData/blob/55dc8bda8a98c5de36ce738b0af2bb67748943a6/src/Microsoft.AspNetCore.OData/Extensions/LinkGeneratorHelpers.cs#L37.. Hi. you can try these HttpRequest extension methods to create a URL that you'd like to add as a metadata reference. Let's know if you are able to do it or if you have any questions.

**Reply by perci994** [reply]
Hi, thank you for help. I tried use this method but im not sure if i use it correctly.

```
public virtual OutputDto Handle(ODataQueryOptions<AllBomItemOutputDto> query)
{
	.....

       var url = query.Request.CreateODataLink();
           
       return result;
 }
```

Url returns me this value: "https://localhost:5001/". Do i something wrong?
Thank you

**Reply by xuzhg** [reply]
Try to add parameters when calling `query.Request.CreateODataLink()`.

query.Request.CreateODataLink(MetadataSegemnt.Instance, ....)

**Reply by petrkasnal** [reply]
@xuzhg  Thank you for the answer. And where exactly do I get the MetadataSegemnt.Instance?

**Reply by xuzhg** [reply]
Sorry, there's a typo. It should be "MetadataSegment". 

```C#
namespace Microsoft.OData.UriParser;

//
// Summary:
//     A segment representing $metadata in a path.
public sealed class MetadataSegment : ODataPathSegment
```

You can use the following sample to get a Uri:

<img width="656" alt="image" src="https://github.com/OData/AspNetCoreOData/assets/9426627/f9bcb323-5e73-4e00-a5b6-fcd21eb1bd8a">

So, if you need '#Order', you can append it by yourself.

**Reply by petrkasnal** [reply]
But i don't want create it manually. I hope that can be create autoamtically like normaly OData do it.
